language/tools/move-package/src/resolution/dependency_graph.rs

```diff
@@ -415,45 +491,106 @@ impl DependencyGraph {
+        ext_graph: &DiGraphMap<PM::PackageName, Dependency>,
+        ext_table: &BTreeMap<PM::PackageName, Package>,
+        resolver: Symbol,
+        overrides: &BTreeMap<PM::PackageName, Package>,
+    ) -> Result<()> {
+        ext_pkg.resolver = Some(resolver);
+
+        // The root package is not present in the package table (because it doesn't have a
+        // source).  If it appears in the other table, it indicates a cycle.
+        if ext_name == self.root_package {
+            bail!(
+                "Conflicting dependencies found:\n{0} = 'root'\n{0} = {1}",
+                ext_name,
+                PackageWithResolverTOML(&ext_pkg),
+            );
+        }
 
-                // Seeing the same package in `extension` is OK only if it has the same set of
-                // dependencies as the existing one.i
-                Entry::Occupied(_) => {
-                    let (self_deps, ext_deps) =
-                        pkg_deps_equal(ext_name, &self.package_graph, &ext_graph);
-                    if self_deps != ext_deps {
-                        bail!(
+        match self.package_table.entry(ext_name) {
+            Entry::Vacant(entry) => {
+                entry.insert(ext_pkg);
+            }
+
+            // Seeing the same package in `extension` is OK only if it has the same set of
+            // dependencies as the existing one.i
+            Entry::Occupied(entry) if entry.get() == &ext_pkg => {
+                let (self_deps, ext_deps) =
+                    pkg_deps_equal(ext_name, &self.package_graph, ext_graph);
+                if self_deps != ext_deps {
+                    bail!(
                             "Conflicting dependencies found for '{ext_name}' during external resolution by '{resolver}':\n{}{}",
                             format_deps("\nExternal dependencies not found:", self_deps),
                             format_deps("\nNew external dependencies:", ext_deps),
                         );
-                    }
                 }
+                // check if acyclic to avoid infinite recursion
+                self.check_acyclic()?;
+                self.override_verify(&ext_pkg, ext_name, overrides)?;
+                // same package, no reason to process its dependencies
+                return Ok(());
             }
-        }
 
-        // Because all the packages in `extensions`'s package table didn't exist in `self`'s, all
-        // `ext_graph`'s edges are known to not occur in `self.package_graph` and can be added
-        // without worrying about introducing duplicate edges.
-        for (from, to, dep) in ext_graph.all_edges() {
-            self.package_graph.add_edge(from, to, dep.clone());
+            Entry::Occupied(mut entry) => {
+                if let Some(overridden_pkg) = overrides.get(&ext_name) {
+                    // override found - use it and return - no reason to process its dependencies
+                    entry.insert(overridden_pkg.clone());
```

Comment:
Do we need to do the acyclicity check here?

Reply:
Good point. I assumed we might as I don't think we have any guards against actively creating cycles during graph construction nor does `override_verify` include such check when traversing the graph, which could lead to infinitely traversing the cycle.

I will try creating an example of such behavior to either confirm the use of  acyclicity here or rule it out.